argo_jupyter_scheduler/executor.py

```diff
@@ -369,7 +383,24 @@ def _create_cwf_oject(
                         "job_definition_id": job_definition_id,
                     },
                 )
+
                 Step(name="main", template=main, continue_on=ContinueOn(failed=True))
+
+                token, channel = get_slack_token_channel(envs)
+                if token is not None and channel is not None:
+                    send_to_slack(
+                        name="send-to-slack",
+                        arguments={
+                            "token": token,
+                            "channel": channel,
+                            "file_path": str(gen_html_path(input_path)),
+                        },
+                        when=successful,
+                        continue_on=ContinueOn(failed=True),
+                    )
+                    failure += " || {{steps.send-to-slack.status}} == Failed"
+                    successful += " && {{steps.send-to-slack.status}} == Succeeded"
```

Comment:
also, instead of string manipulation, we could do something like:

(not necessarily in this PR though, writing it for future)

```python
class JobState:

    def __init__(self):
        self.failure_conditions = []
        self.successful_conditions = []

    def add_failure(self, status_param):
        self.failure_conditions.append("{{ " + status_param + " }}")

    def add_success(self, status_param):
        self.successful_conditions.append("{{ " + status_param + " }}")

    @property
    def failure(self):
        return " || ".join(self.failure_conditions)

    @property
    def success(self):
        return " && ".join(self.successful_conditions)

    def update(self, db_url, job_definition_id):
        update_job_status_failure(
            name="failure",
            arguments={
                "db_url": db_url,
                "job_definition_id": job_definition_id,
            },
            when=self.failure,
        )

        update_job_status_success(
            name="success",
            arguments={
                "db_url": db_url,
                "job_definition_id": job_definition_id,
            },
            when=self.success,
        )

```

Then we could simply do, something like:

```python
job_state = JobState()
job_state.add_failure("steps.send-to-slack.status")
job_state.add_success("steps.send-to-slack.status")
job_state.update(db_url, job_definition_id)
```

Reply:
Yeah, I dislike how everything is repeated twice. But not going to change as part of this as I want to make minimal changes.

Reply:
yep, makes sense.

Reply:
Added an issue so we don't forget: https://github.com/nebari-dev/argo-jupyter-scheduler/issues/5